bluemira/geometry/coordinates.py

```diff
@@ -0,0 +1,943 @@
+
+    Parameters
+    ----------
+    x: np.array
+        x coordinates of the loop to calculate on
+    y: np.array
+        y coordinates of the loop to calculate on
+    z: Union[None, np.array]
+
+    Returns
+    -------
+    centroid: np.array
+        The x, y, [z] coordinates of the centroid [m]
+    """
+    if z is None:
+        return get_centroid_2d(x, y)
+    else:
+        return get_centroid_3d(x, y, z)
+
+
+@nb.jit(cache=True, nopython=True)
+def get_centroid_2d(x, z):
+    """
+    Calculate the centroid of a non-self-intersecting 2-D counter-clockwise polygon.
+
+    Parameters
+    ----------
+    x: np.array
+        x coordinates of the loop to calculate on
+    z: np.array
+        z coordinates of the loop to calculate on
+
+    Returns
+    -------
+    centroid: List[float]
+        The x, z coordinates of the centroid [m]
+    """
+    if not check_ccw(x, z):
+        x = x[::-1]
+        z = z[::-1]
+    area = get_area_2d(x, z)
+
+    cx, cz = 0, 0
+    for i in range(len(x) - 1):
+        a = x[i] * z[i + 1] - x[i + 1] * z[i]
+        cx += (x[i] + x[i + 1]) * a
+        cz += (z[i] + z[i + 1]) * a
+
+    if area != 0:
+        # Zero division protection
+        cx /= 6 * area
+        cz /= 6 * area
+
+    return [cx, cz]
+
+
+def get_centroid_3d(x, y, z):
+    """
+    Calculate the centroid of a non-self-intersecting counterclockwise polygon
+    in 3-D.
+
+    Parameters
+    ----------
+    x: Iterable
+        The x coordinates
+    y: Iterable
+        The y coordinates
+    z: Iterable
+        The z coordinates
+
+    Returns
+    -------
+    centroid: List[float]
+        The x, y, z coordinates of the centroid [m]
+    """
+    cx, cy = get_centroid_2d(x, y)
+    cx2, cz = get_centroid_2d(x, z)
+    cy2, cz2 = get_centroid_2d(y, z)
+
+    # The following is an "elegant" but computationally more expensive way of
+    # dealing with the 0-area edge cases
+    # (of which there are more than you think)
+    cx = np.array([cx, cx2])
+    cy = np.array([cy, cy2])
+    cz = np.array([cz, cz2])
+
+    def get_rational(i, array):
+        """
+        Gets rid of infinity and nan coordinates
+        """
+        args = np.argwhere(np.isfinite(array))
+        if len(args) == 0:
+            # 2-D shape with a simple axis offset
+            # Get the first value of the coordinate set which is equal to the
+            # offset
+            return [x, y, z][i][0]
+        elif len(args) == 1:
+            return array[args[0][0]]
+        else:
+            if not np.isclose(array[0], array[1]):
+                # Occasionally the two c values are not the same, and one is 0
+                # Take non-trivial value (this works in the case of 2 zeros)
+                return array[np.argmax(np.abs(array))]
+            else:
+                return array[0]
+
+    return [get_rational(i, c) for i, c in enumerate([cx, cy, cz])]
+
+
+def rotation_matrix_v1v2(v1, v2):
```

Comment:
SonarCloud also sees this function duplicated in https://github.com/Fusion-Power-Plant-Framework/bluemira/blob/9e94a3f4e58d73bab1521a0b5c6468b0333de139/BLUEPRINT/geometry/geombase.py#L136 and https://github.com/Fusion-Power-Plant-Framework/bluemira/blob/9e94a3f4e58d73bab1521a0b5c6468b0333de139/BLUEPRINT/utilities/plottools.py#L263. Can we use this function in the implementation of those two static functions?